runtime/lua/vim/lsp/document_color.lua

```diff
@@ -0,0 +1,351 @@
+end
+
+--- Cache of the highlight groups that we've already created.
+--- @type table<string, true>
+local color_cache = {}
+
+--- Gets or creates the highlight group for the given LSP color information.
+---
+--- @param hex_code string
+--- @param style string
+--- @return string
+local function get_hl_group(hex_code, style)
+  if style ~= 'background' then
+    style = 'foreground'
+  end
+
+  local hl_name = ('LspDocumentColor_%s_%s'):format(hex_code:sub(2), style)
+
+  if not color_cache[hl_name] then
+    if style == 'background' then
+      api.nvim_set_hl(0, hl_name, { bg = hex_code, fg = get_contrast_color(hex_code) })
+    else
+      api.nvim_set_hl(0, hl_name, { fg = hex_code })
+    end
+
+    color_cache[hl_name] = true
+  end
+
+  return hl_name
+end
+
+--- |lsp-handler| for the `textDocument/documentColor` method.
+---
+--- @param err? lsp.ResponseError
+--- @param result? lsp.ColorInformation[]
+--- @param ctx lsp.HandlerContext
+local function on_document_color(err, result, ctx)
+  if err then
+    lsp.log.error('document_color', err)
+    return
+  end
+
+  local bufnr = assert(ctx.bufnr)
+  local bufstate = bufstates[bufnr]
+  local client_id = ctx.client_id
+
+  if
+    util.buf_versions[bufnr] ~= ctx.version
+    or not result
+    or not api.nvim_buf_is_loaded(bufnr)
+    or not bufstate.enabled
+  then
+    return
+  end
+
+  if not bufstate.hl_info or not bufstate.buf_version then
+    bufstate.hl_info = {}
+    bufstate.buf_version = ctx.version
+  end
+
+  local hl_infos = {} --- @type vim.lsp.document_color.HighlightInfo[]
+  local style = document_color_opts.style
+  for _, res in ipairs(result) do
+    local range = {
+      res.range.start.line,
+      res.range.start.character,
+      res.range['end'].line,
+      res.range['end'].character,
+    }
+    local hex_code = get_hex_code(res.color)
+    local hl_info = { range = range, hex_code = hex_code }
+
+    if type(style) == 'string' then
+      hl_info.hl_group = get_hl_group(hex_code, style)
+    end
+
+    table.insert(hl_infos, hl_info)
+  end
+  bufstate.hl_info[client_id] = hl_infos
+
+  bufstate.buf_version = ctx.version
+  api.nvim__redraw({ buf = bufnr, valid = true, flush = false })
+end
+
+--- @param bufnr integer
+local function buf_clear(bufnr)
+  local bufstate = bufstates[bufnr]
+  local client_ids = bufstate and vim.tbl_keys(bufstate.hl_info) or {} --- @type integer[]
+
+  for _, id in ipairs(client_ids) do
+    bufstate.hl_info[id] = {}
+  end
+
+  api.nvim_buf_clear_namespace(bufnr, document_color_ns, 0, -1)
+  api.nvim__redraw({ buf = bufnr, valid = true, flush = false })
+end
+
+--- @param bufnr integer
+--- @param client_id? integer
+local function buf_refresh(bufnr, client_id)
+  util._refresh(ms.textDocument_documentColor, {
+    bufnr = bufnr,
+    handler = on_document_color,
+    client_id = client_id,
+  })
+end
+
+--- @param bufnr integer
+local function buf_disable(bufnr)
+  buf_clear(bufnr)
+
+  bufstates[bufnr] = { enabled = false }
```

Comment:
```suggestion
  bufstates[bufnr] = { enabled = false, applied_version = {}, hl_info = {} }
```

If we make the fields required, `lua_ls` will detect insufficient field initialization at every initialization. This way, we can ensure that both `bufstate` and the nested tables inside it exist simultaneously.